src/linter/block.go

```diff
@@ -2302,7 +2315,27 @@ func (b *BlockWalker) handleAssign(a *assign.Assign) bool {
 		b.checkVoidType(a.Expression)
 		b.replaceVar(v, solver.ExprTypeLocal(b.ctx.sc, b.r.ctx.st, a.Expression), "assign", meta.VarAlwaysDefined)
 	case *expr.List:
-		b.handleAssignList(v.Items)
+		if !meta.IsIndexingComplete() {
+			return true
+		}
+
+		tp := solver.ExprType(b.ctx.sc, b.r.ctx.st, a.Expression)
+		var shapeType string
+		tp.Iterate(func(t string) {
+			if strings.HasPrefix(t, `\shape`) {
```

Comment:
Not related to a data race, but it should be `\shape$` to avoid collisions with `\shape` class (if user defined it). Auto-generated shape types have `$` in their name.